### PR TITLE
Update npm to Bootstrap 4 (#2807)

### PR DESCRIPTION
**Pull request by tywrenn**
This is in reference to #2807 to update to Bootstrap 4. We can have Bootstrap 3 on if we want to, but since we are probably gonna go all in to update to Bootstrap 4 I thought why even bother keeping it.

This is a bleeding edge commit and definitely will cause internal issues with OpenEMR. It is not recommended to update to this commit until the framework is fully fixed.

To test out the new features coming from this pull request, a demo is available on the OpenEMR website: https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

### CODE REVIEW

**Comment by tywrenn** [comment]
Just realized I accidentally put a comma in the json that makes it fail - fixing this now

**Comment by tywrenn** [comment]
Ok after building in docker compose it shows that brace-expansion will need to update to bootstrap 4 as well @sjpadgett i'm not sure what that dependency does. Also, i18next gets auto-updated in a docker compose environment for some reason and adds the babel runtime as a dependency (guess this was force-updated by npm?)

**Comment by tywrenn** [comment]
*correction emodal

**Comment by sjpadgett** [comment]
emodal needs to go away and will be something I need to do. Used in portal and institutional billing so just work around for now.

**Comment by tywrenn** [comment]
This updates us to Bootstrap 4 now - I triple checked this time

**Comment by bradymiller** [comment]
Up For Grabs demo for this PR is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

**Comment by stephenwaite** [comment]
nice
looks like the patient menu navbar is out of whack; remember navbars being discussed so probably not surprising

**Comment by tywrenn** [comment]
@stephenwaite Yeah it appears as though nav-pills is not working because navbars require nav-item classes as well as nav-link classes. Also some other menus are bound to not work due to spacing issues.

**Comment by zbig01** [comment]
@tywrenn 
Thanks for getting Bootstrap 4 in.

FYI several features that worked in Bootstrap 3 are now broken. Several responsive forms that were working now are displaying differently. Have to look at some examples.

**EOB-BS3**

EOB_BS3

**EOB-BS4**

EOB_BS4

**Messages BS3**

Messages_BS3

**Messages BS4**

Messages_BS4

**Procedures BS3**

Procedures_BS3

**Procedures BS4**

Procedures_BS4

As you are well aware there are some critical differences in font sizes for headings in BS 3 vs BS 4. The headings sizes in BS 3 were slightly smaller and was working well. I need to look at the other classes more closely to see the differences. 

To make all this work in both BS 3 and BS 4 would require some thought.

Cheers and once again a great effort

**Comment by sjpadgett** [comment]
BS4 All the way. Trying to run BS4 and BS3 is just more work.

**Comment by matrix-amiel** [comment]
Hi @sjpadgett 

We tested this branch and we found bugs with LTR languages-
image
And more with RTL-
image

**Comment by sjpadgett** [comment]
Hi @matrix-amiel Yep we know BS4 is not ready. What I was hoping for from you guys is maybe you have looked over the BS4-RTL library and formed an opinion on it's inclusion in OpenEMR.

Your group did a lot of the BS3 RTL and any advise you can give us here is appreciated.

**Comment by tywrenn** [comment]
This commit is loaded with bugs so more testing is needed
- Overhauls patient portal (needs further changes)
- Fixed login page logo
- replaces pull-left and pull-right with new classes
- Other misc. Fixes
php sniffer isn't happy with me 😂

**Comment by tywrenn** [comment]
Fixed Travis CI

**Comment by tywrenn** [comment]
This commit replaces the outdated `col-xs-*` with the new `col-*` as well as replaces `btn-default` with `btn-secondary` as well as other code fixes.

**Comment by sjpadgett** [comment]
Hi @tywrenn I'd highly recommend that with any new merged PR's, you try to pull into your local branch and fix any conflicts as you go or we'll have a nightmare on our hands when time to pull into master.

**Comment by tywrenn** [comment]
This commit does the following:
- Replaces outdated `col-xs-*` classes with new `col-*` classes
- Replaces `table-condensed` with new `table-sm`
- Removes outdated `col-*-offset-*` class format with new `offset-*-*` class format
- Replaces `btn-xs` class with `btn-sm` (needs community testing)
- Fixes Flow Board from not displaying correctly (this code was actually coded incorrectly!)
- Replaces `hidden-*` classes with `d-*-none` classes
- Fixes FontAwesome size issue where icons were becoming large
- Working on fixing jQuery issue with patient portal (yet now fixed navigating between the tabs)
- Updates to latest Bootstrap 4.4.0
- Other bug fixes

**Comment by sjpadgett** [comment]
@tywrenn Need to give us permissions to push to your branch. Otherwise, nobody can help leaving you by your wee self :)...

**Comment by tywrenn** [comment]
@sjpadgett Invited you :)

**Comment by bradymiller** [comment]
Very exciting stuff! Will hold off an bringing in any PRs into master unless ok with @sjpadgett (to ensure not stepping on toes). Are there any specific issues/problems/bugs that other can work on that will help out?

File: interface/super/edit_globals.php [hunk 1]
```diff
@@ -435,7 +441,7 @@ function validate_file() {
                         ?>
                           <input name='srch_desc' id='srch_desc' class='form-control' type='text' placeholder='<?php echo $placeholder; ?>' value='<?php echo (!empty($_POST['srch_desc']) ? attr($_POST['srch_desc']) : '') ?>' />
                         <span class="input-group-btn">
-                            <button class="btn btn-secondary btn-search" type='submit' id='globals_form_search' name='form_search'><?php echo xlt('Search'); ?></button>
+                            <button class="btn btn-default btn-search" type='submit' id='globals_form_search' name='form_search'><?php echo xlt('Search'); ?></button>
```
Comment (tywrenn) [inline interface/super/edit_globals.php]: Default is gone in Bootstrap 4...

File: interface/super/edit_globals.php [hunk 2]
```diff
@@ -874,13 +880,13 @@ function validate_file() {
     if (winWidth > 1024) {
         if (!userMode) {
             $('.row  .control-label, .row  .oe-input').removeClass('col-sm-6');
-            $('.row  .control-label').addClass('col-sm-4 offset-sm-1');
+            $('.row  .control-label').addClass('col-sm-4 col-sm-offset-1');
```
Comment (tywrenn) [inline interface/super/edit_globals.php]: This doesn't work in Bootstrap 4 - new `offset-sm-1` replaces it.

File: interface/super/edit_globals.php [hunk 3]
```diff
@@ -874,13 +880,13 @@ function validate_file() {
...
             $('.row  .oe-input').addClass('col-sm-4');
         }
     } else {
         if (!userMode) {
             $('.row  .control-label, .row  .oe-input').addClass('col-sm-6');
-            $('.row  .control-label').removeClass('col-sm-4 offset-sm-1');
+            $('.row  .control-label').removeClass('col-sm-4 col-sm-offset-1');
```
Comment (tywrenn) [inline interface/super/edit_globals.php]: Same issue here

File: portal/messaging/messages.php [hunk 4]
```diff
@@ -697,7 +697,7 @@ class="btn btn-primary btn-sm"
                     </div>
                 </div>
                 <!--/inbox panel-->
-                <div class="well well-s text-right">
+                <div class="jumbletron jumbletron-sm well-s text-right">
```
Comment (tywrenn) [inline portal/messaging/messages.php]: Did you mean `jumbotron` here?

File: portal/patient/templates/DefaultErrorFatal.tpl.php [hunk 5]
```diff
@@ -19,13 +19,13 @@
     <!-- this is used by app.js for scraping -->
     <!-- ERROR <?php $this->eprint($this->message); ?> /ERROR -->
 
-    <h3 onclick="$('#stacktrace').show('slow');" class="well" style="cursor: pointer;"><?php $this->eprint($this->message); ?></h3>
+    <h3 onclick="$('#stacktrace').show('slow');" class="jumbletron jumbletron-sm" style="cursor: pointer;"><?php $this->eprint($this->message); ?></h3>
```
Comment (tywrenn) [inline portal/patient/templates/DefaultErrorFatal.tpl.php]: It's `jumbotron`

File: portal/patient/templates/DefaultErrorFatal.tpl.php [hunk 6]
```diff
@@ -19,13 +19,13 @@
...
 
     <p><?php echo xlt('You may want to try returning to the the previous page and verifying that all fields have been filled out correctly.'); ?></p>
 
     <p><?php echo xlt('If you continue to experience this error please contact support.'); ?></p>
 
-    <div id="stacktrace" class="well hide">
+    <div id="stacktrace" class="jumbletron jumbletron-sm hide">
```
Comment (tywrenn) [inline portal/patient/templates/DefaultErrorFatal.tpl.php]: Same issue here

File: portal/patient/templates/DefaultHome.tpl.php [hunk 7]
```diff
@@ -31,7 +31,7 @@
     </div>
 </div>
 <div class="container">
-    <div class='well'>
+    <div class='jumbletron jumbletron-sm'>
```
Comment (tywrenn) [inline portal/patient/templates/DefaultHome.tpl.php]: Same issue here

File: portal/patient/templates/DefaultHome.tpl.php [hunk 8]
```diff
@@ -40,7 +40,7 @@
             data-target="#formdialog" href="#"><?php echo xlt('Tell me more'); ?> »</a>
     </div>
 </div>
-<div class='well'>
+<div class='jumbletron jumbletron-sm'>
```
Comment (tywrenn) [inline portal/patient/templates/DefaultHome.tpl.php]: Same issue here

File: portal/patient/templates/PatientListView.tpl.php [hunk 9]
```diff
@@ -69,7 +69,7 @@
     <!-- Could add/manage table list here -->
 <script type="text/template" id="patientModelTemplate"> <!-- -->
 
-<div id='profileHelp' class='well' style='display:none;width: 650px; margin: 0 auto;'>
+<div id='profileHelp' class='jumbletron jumbletron-sm' style='display:none;width: 650px; margin: 0 auto;'>
```
Comment (tywrenn) [inline portal/patient/templates/PatientListView.tpl.php]: It's `jumbotron`

File: portal/patient/templates/SecureApp.tpl.php [hunk 10]
```diff
@@ -39,7 +39,7 @@
             </p>
         </div>
 
-        <form class="well" method="post" action="login">
+        <form class="jumbletron jumbletron-sm" method="post" action="login">
```
Comment (tywrenn) [inline portal/patient/templates/SecureApp.tpl.php]: Jumbotron here

File: portal/patient/templates/ProviderHome.tpl.php [hunk 11]
```diff
@@ -48,7 +48,7 @@
             data-target="#formdialog" href="#"><?php echo xlt('Tell me more') ?> »</a></p>
     </div>
 </div>
-<div class='well'>
+<div class='jumbletron jumbletron-sm'>
```
Comment (tywrenn) [inline portal/patient/templates/ProviderHome.tpl.php]: Jumbotron here

File: portal/patient/templates/ProviderHome.tpl.php [hunk 12]
```diff
@@ -38,7 +38,7 @@
     </div>
 </div>
 <div class="container bg-info">
-    <div class='well'>
+    <div class='jumbletron jumbletron-sm'>
```
Comment (tywrenn) [inline portal/patient/templates/ProviderHome.tpl.php]: Jumbotron here

**Comment by sjpadgett** [comment]
lol, could of just said "hey dummy! You misspelt jumbotron"...
I'll fix. Also going to push portal patient document to you soon. I've always have had problem deciding how workflow should be and I like how I have it in general. Need someone with keener BS4 eye than eye to check me on the menus and improve.

File: interface/super/edit_globals.php (hunk 3 above)
Comment (sjpadgett) [reply interface/super/edit_globals.php]: I didn't change these... branch may have been outdated

File: portal/patient/templates/DefaultHome.tpl.php [hunk 13]
```diff
@@ -31,7 +31,7 @@
     </div>
 </div>
 <div class="container">
-    <div class='jumbletron jumbletron-sm'>
+    <div class='jumbotron jumbotron-fluid'>
     <div class="jumbotron">
```
Comment (tywrenn) [inline portal/patient/templates/DefaultHome.tpl.php]: I'll fix this issue right here.

**Comment by bradymiller** [comment]
btw, disregard the travis issues that will arise in future commits. New version of squizlabs/php_codesniffer is catching unrelated issues. We will fix them after this PR gets into the codebase (this will prevent merge suicide by @sjpadgett :) ).

File: portal/patient/templates/DefaultHome.tpl.php (hunk 13 above)
Comment (sjpadgett) [reply portal/patient/templates/DefaultHome.tpl.php]: originally wanted a jumbotron-sm. Where can I work with best bet at staying out of your way due to us merging in your branch? Is your next commit going to include many files?

**Comment by bradymiller** [comment]
i think i answered my own question. just tested the up for grbas demo and looks like the current themes are all still working.

**Comment by bradymiller** [comment]
On testing, noted that the use of the bootstrap-sidebar package is breaking at Administration->Practice->Practice Settings.
This is not surprising since that package is specific to bootstrap 3 and is dead (no commits to it in 5 years).
The good news is that it is not used anywhere else in openemr :)

**Comment by bradymiller** [comment]
Also, due to a github PR glitch, tough for me to review this code (github PR diff's get messed up when original commits from master are missing (since Jerry's commit basically squashed them into one; not an issue outside github, but for some reason github PR diff can't deal with them)). So, at some point (maybe when @sjpadgett and @tywrenn need to come up for air :) ), will be good to bring in another PR to fix this (let me know when is a good time and I can create the branch that @tywrenn 
 can then grab and push another PR with).

**Comment by sjpadgett** [comment]
I'll pull the package and replace template as soon as we get our side bars agreeable. Pretty close in patient doc in portal. Need to get auto sizing yet.

**Comment by sjpadgett** [comment]
> Also, due to a github PR glitch

Tyler may be at school or work now but, I have admin on his branch and nothing is staged there now so we can do now if you want. What do I need to do?

**Comment by bradymiller** [comment]
Note it will involve a forced push, and the fact that this is on tywrenn's master branch makes me very wary to do this (if we mistakenly force push to official master then we break all mirrored repos and lose 1000 forks; ie. complete and utter disaster). Much better to make another branch on tywrenn's repo that is NOT master and then push it to make another PR. Here is the branch that can be used for that:
https://github.com/bradymiller/openemr/commits/tywrenn-bootstrap-rebased-branch

Can collect the branch to your repo via:
git repo add bradymiller git@github.com:bradymiller/openemr.git
git fetch bradymiller
git checkout bradymiller/tywrenn-bootstrap-rebased-branch
git checkout -b bootstrap-updates

Now you have a bootstrap-updates branch in your repo that you (or @tywrenn) can push to a PR.

**Comment by tywrenn** [comment]
Or I can just give you collab access to my repo @bradymiller and let you do the work 😂

**Comment by bradymiller** [comment]
Note I corrected a mistake in above commands to (changed repo to remote):
git remote add bradymiller git@github.com:bradymiller/openemr.git

**Comment by bradymiller** [comment]
@tywrenn ,
Sounds good.

**Comment by tywrenn** [comment]
@bradymiller just invited you

**Comment by bradymiller** [comment]
Actually, @tywrenn , I still can't bring in a PR in your name. Are you able to do the following:
```
git remote add bradymiller git@github.com:bradymiller/openemr.git
git fetch bradymiller
git checkout bradymiller/tywrenn-bootstrap-rebased-branch
git checkout -b bootstrap-updates
```
The create a PR with your new shiny bootstrap-updates branch?

**Comment by sjpadgett** [comment]
@tywrenn You should always allow admin access on PR's so we can mess you up if we get mad at ya :)

**Comment by bradymiller** [comment]
btw, sorry in advance for the 1000+ branches that you will collect when fetch bradymiller :)

**Comment by tywrenn** [comment]
[branchesohmy.txt](https://github.com/openemr/openemr/files/3923929/branchesohmy.txt)
God dang Brady what have you done 🤣

**Comment by tywrenn** [comment]
Thought rebasing doesn't make us have to recreate this pull request...@bradymiller

**Comment by bradymiller** [comment]
@tywrenn , yep, github even flagged me as a bot a couple years back :)
Trying to salvage this PR is a lost cause (force push/rebasing a branch causes headaches to folks working on the same branch).
Recommend a new PR (I'll update the demo and provide link backs on old/new PR). And we'll avoid this in future by doing merges from upstream/master when needed (I am super happy to do this kind of thing). Planning to actually bring in some ongoing work when completed (https://github.com/openemr/openemr/pull/2830; also gonna do the zend js/css asset updates in that PR) into this PR instead of master since more related to this project (when ready I'll package it into 1 commit and ensure you or Jerry have no ongoing work pending commit).

**Comment by bradymiller** [comment]
Placed a new PR here from @tywrenn 's repo (now that I have all encompassing access to @tywrenn's repo (play evil laugh soundtrack :) ) ) so then tywrenn has main control of the branch:
https://github.com/openemr/openemr/pull/2832

I'll transfer over the up for grabs demo to that and will keep it updated intermittently via merges from upstream/master.

Gonna close this PR

**Pull request closed without merging**